### PR TITLE
added denoising code

### PR DESCRIPTION
**Pull request by nd2506**

### CODE REVIEW

File: src/yass/cluster/cluster.py [hunk 1]
```diff
@@ -376,8 +377,8 @@ def initialize(self, initial_spt):
         if self.plotting:
             self.x = np.zeros(100, dtype = int)
             self.fig1 = plt.figure(figsize =(60,60))
-            self.grid1 = plt.GridSpec(20,20,wspace = 0.0,hspace = 0.2)
-            self.ax1 = self.fig1.add_subplot(self.grid1[:,:])
+            self.grid1 = plt.GridSpec(20,20,wspace = 1,hspace = 2)
+
```
Comment (jinhyunglee) [inline src/yass/cluster/cluster.py]: is it okay without `self.ax1`?

Comment (nd2506) [reply src/yass/cluster/cluster.py]: yes, it just removes the outside box.

File: src/yass/cluster/util.py [hunk 2]
```diff
@@ -2366,7 +2367,123 @@ def connected_channels(channel_list, ref_channel, neighbors, keep=None):
...
+    
+    if ref is None:
+        ref_upsampled = wf_mc_up.mean(0)
+    else:
+        ref_upsampled = signal.resample(ref, wlen)
+        
+    ref_shifted = np.zeros([wlen, nshifts])
+
+    for i,s in enumerate(range(-int((nshifts-1)/2), int((nshifts-1)/2+1))):
+        ref_shifted[:,i] = np.roll(ref_upsampled, -s)
```
Comment (jinhyunglee) [inline src/yass/cluster/util.py]: we decided not to use np.roll..
I thought that you wrote a function that does not do np.roll?

File: src/yass/cluster/util.py [hunk 3]
```diff
@@ -2480,17 +2598,21 @@ def global_merge_max_dist(chunk_dir, CONFIG, out_dir, units):
...
+#             spike_width = templates.shape[1]
+#             temps_stack = np.swapaxes(templates, 1,0)
+#             temps_stack = np.reshape(temps_stack, (temps_stack.shape[0], 
+#                                      temps_stack.shape[1]*temps_stack.shape[2])).T
+#             _, temps_PCA, pca_object = PCA(temps_stack, n_pca)
+
+#             temps_PCA = temps_PCA.reshape(templates.shape[0],templates.shape[2],templates.shape[1])
+#             temps_PCA = np.swapaxes(temps_PCA,1,2)
+            pca_mc = pickle.load(open('/hdd/data/Nishchal/yass/src/pca_mc.p', 'rb'))
+            pca_sec = pickle.load(open('/hdd/data/Nishchal/yass/src/pca_sec.p', 'rb'))
```
Comment (jinhyunglee) [inline src/yass/cluster/util.py]: You said you tried denoising using pca_mc only and said it is working fine. 
Did you find failure cases later?

File: src/yass/cluster/util.py (hunk 2 above)
Comment (nd2506) [reply src/yass/cluster/util.py]: I have that too. but that cuts the wf by half. Do you want that version? what's wrong with np.roll?

File: src/yass/cluster/util.py (hunk 3 above)
Comment (nd2506) [reply src/yass/cluster/util.py]: No I did not find failure cases. But I included both just for completeness

File: src/yass/cluster/util.py (hunk 2 above)
Comment (jinhyunglee) [reply src/yass/cluster/util.py]: check the documentation of np.roll to see what it does. This function is not about shifting, especially if you do it for larger shifts.

File: src/yass/cluster/util.py (hunk 3 above)
Comment (jinhyunglee) [reply src/yass/cluster/util.py]: I don't understand it.
I thought that your conclusion was using pca_mc is better on secondary channels. But now you are now using pca_sec to denoise.

Comment (jinhyunglee) [reply src/yass/cluster/util.py]: Of course I have to read the code in detail..
now it is using pca_sec to denoise secondary channels. 
I will keep it as it is.

**Pull request merged**